File: backend/src/test/scala/com/skewwhiffy/auraltester/internalnotation/InternalNotationFactoryTest.scala
package com.skewwhiffy.auraltester.internalnotation

import com.skewwhiffy.auraltester.clefs.ClefFactory
import com.skewwhiffy.auraltester.testutils.{MockInstantiation, TestData}
import org.mockito.{InjectMocks, Mock}
import org.scalatest.flatspec.AnyFlatSpec
import org.scalatest.matchers.should

class InternalNotationFactoryTest extends AnyFlatSpec with should.Matchers with MockInstantiation {
  @Mock
  private val clefFactory: ClefFactory = null
  @Mock
  private val intervalFactory: IntervalFactory = null
  @Mock
  private val keyFactory: KeyFactory = null
  @Mock
  private val noteFactory: NoteFactory = null
  @InjectMocks
  private val internalNotationFactory: InternalNotationFactory = null

  List("treble", "alto", "tenor", "bass").foreach(it => {
    it should s"when $it then proxies to clefFactory" in {
      val expected = TestData.random.clef
      val method = classOf[ClefFactory].getMethod(it)
      when(method.invoke(clefFactory)).thenReturn(expected)

      val actual = internalNotationFactory.clef(it)

      assert(actual == expected)
    }
  })

  it should "when clef name invalid then blows up" in {
    assertThrows[IllegalArgumentException] {
      internalNotationFactory.clef("not a clef name")
    }
  }

  it should "when getNote then proxies to noteFactory" in {
    val rawNote = TestData.random.string
    val expected = TestData.random.absoluteNote
    when(noteFactory.getAbsoluteNote(rawNote)).thenReturn(expected)

    val actual = internalNotationFactory.getNote(rawNote)

    assert(actual == expected)
  }

  it should "when getNotes then proxies to noteFactory" in {
    val rawNotes = Range(0, 10).map(_ => TestData.random.string)
    val expected = rawNotes.map(_ => TestData.random.absoluteNote)
    rawNotes.zip(expected).foreach(it => {
      when(noteFactory.getAbsoluteNote(it._1)).thenReturn(it._2)
    })

    val actual = internalNotationFactory.getNotes(rawNotes.mkString(" "))

    assert(actual == expected)
  }

  it should "when getDirectedInterval then proxies to intervalFactory" in {
    val rawInterval = TestData.random.string
    val expected = TestData.random.directedInterval
    when(intervalFactory.getDirectedInterval(rawInterval)).thenReturn(expected)

    val actual = internalNotationFactory.getDirectedInterval(rawInterval)

    assert(actual == expected)
  }

  it should "when getKey then proxies to keyFactory" in {
    val rawKey = TestData.random.string
    val expected = TestData.random.key
    when(keyFactory.getKey(rawKey)).thenReturn(expected)

    val actual = internalNotationFactory.getKey(rawKey)

    assert(actual == expected)
  }
}

File: backend.kotlin/src/test/kotlin/com/skewwhiffy/notation/factory/InternalNotationFactoryTest.kt
package com.skewwhiffy.notation.factory

import com.skewwhiffy.exception.UnrecognizedClefException
import com.skewwhiffy.test.util.TestData
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.fail
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.api.extension.ExtendWith
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.junit.jupiter.MockitoExtension

@ExtendWith(MockitoExtension::class)
class InternalNotationFactoryTest {
  @Mock
  private lateinit var clefFactory: ClefFactory

  @Mock
  private lateinit var intervalFactory: IntervalFactory

  @Mock
  private lateinit var keyFactory: KeyFactory

  @Mock
  private lateinit var noteFactory: NoteFactory

  @InjectMocks
  private lateinit var internalNotationFactory: InternalNotationFactory

  @ParameterizedTest
  @ValueSource(strings = ["treble", "alto", "tenor", "bass"])
  fun `proxy to clefFactory when clef is requested`(clef: String) {
    val expected = TestData.random.clef
    val method = ClefFactory::class.members.find { it.name == clef }
    method?.let { `when`(it.call(clefFactory)).thenReturn(expected) }
      ?: fail("Could not find member '$clef' on ClefFactory")

    val actual = internalNotationFactory.clef(clef)

    assertThat(actual).isEqualTo(expected)
  }

  @Test
  fun `throw when clef name invalid`() {
    assertThrows<UnrecognizedClefException> { internalNotationFactory.clef("not a clef name") }
  }

  @Test
  fun `proxy to noteFactory when note is requested`() {
    val rawNote = TestData.random.string
    val expected = TestData.random.absoluteNote
    `when`(noteFactory.getAbsoluteNote(rawNote)).thenReturn(expected)

    val actual = internalNotationFactory.getNote(rawNote)

    assertThat(actual).isEqualTo(expected)
  }

  @Test
  fun `proxy to noteFactory when notes are requested`() {
    val rawNotes = (0..10).map { TestData.random.string }
    val expected = rawNotes.map { TestData.random.absoluteNote }
    rawNotes.zip(expected).forEach {
      `when`(noteFactory.getAbsoluteNote(it.first)).thenReturn(it.second)
    }

    val actual = internalNotationFactory.getNotes(rawNotes.joinToString(" "))

    assertThat(actual).isEqualTo(expected)
  }

  @Test
  fun `proxy to intervalFactory when directed interval is requested`() {
    val rawInterval = TestData.random.string
    val expected = TestData.random.directedInterval
    `when`(intervalFactory.getDirectedInterval(rawInterval)).thenReturn(expected)

    val actual = internalNotationFactory.getDirectedInterval(rawInterval)

    assertThat(actual).isEqualTo(expected)
  }


  @Test
  fun `proxy to intervalFactory when directed intervals are requested`() {
    val rawInterval = TestData.random.string
    val expected = listOf(TestData.random.directedInterval)
    `when`(intervalFactory.getDirectedIntervals(rawInterval)).thenReturn(expected)

    val actual = internalNotationFactory.getDirectedIntervals(rawInterval)

    assertThat(actual).isEqualTo(expected)
  }

  @Test
  fun `proxy to keyFactory when key is requested`() {
    val rawKey = TestData.random.string
    val expected = TestData.random.key
    `when`(keyFactory.getKey(rawKey)).thenReturn(expected)

    val actual = internalNotationFactory.getKey(rawKey)

    assertThat(actual).isEqualTo(expected)
  }
}